Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,132 @@
+
+
+## 4. Goals
+
+Provide an MVP version of the DaemonSet deployment of Prometheus Agent to the audience.
+In specific, the MVP will need to:
+* Allow users to deploy one Prometheus Agent pod per node
+* Allow users to restrict on which set of nodes they want to deploy Prometheus Agent
+* Allow each Prometheus Agent pod to scrape from the pods on the same node (PodMonitor support)
+
+
+## 5. Non-Goals
+
+The non-goals are the features that are not easy to implement and require more investigation. We will need to investigate whether there are actual user needs for them, if yes, then how to best implement them. We’ll handle these after the MVP.
+* ServiceMonitor support: There's a performance issue regarding this feature. Since each Prometheus Agent running on a node requires one watch, making all Prometheus Agent pods watch all endpoints will put a huge stress on Kubernetes API server. This is the main reason why GMP hasn’t supported this, even though there are user needs stated in some issues ([#362](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/362), [#192](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192)). However, as discussed with Danny from GMP [here](https://github.com/GoogleCloudPlatform/prometheus-engine/issues/192#issuecomment-2028850846), ServiceMonitor support based on EndpointSlice seems like a viable approach. We’ll investigate this further after the MVP.
+* Storage: We will need to spend time studying more about the WAL, different storage solutions provided by Kubernetes, and how to gracefully handle storage in different cases of crashes. For example, there’s an [issue in Prometheus](https://github.com/prometheus/prometheus/issues/8809) showing that samples may be lost if remote write didn’t flush cleanly. We’ll investigate these further after the MVP.
+* Secret management: We’ll need to figure out how to make each Prometheus Agent pod only watch the secrets from the pods on the same node. This is for both enhancing security and reducing load on Kubernetes API server. However, this is not easy to do. So we’ll handle this after the MVP.
+
+
+## 6. How
+
+### 6.1. CRD:
+
+Currently, we already have a Prometheus Agent CRD that supports StatefulSet deployment. We’ll add new field(s) in this CRD to enable DaemonSet deployment.
+
+The reason for enhancing existing CRD (instead of introducing a new CRD) is it would take less time to finish the MVP. We’ll let users experiment with the MVP, and in case users report a separate CRD is needed, we’ll separate the logic of DaemonSet deployment into a new CRD later.
+
+The current [Prometheus Agent CRD](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1alpha1.PrometheusAgent) already has sufficient fields for the DaemonSet deployment. The DaemonSet deployment can use all the existing fields in the CRD except the ones related to:
+* Selectors for service, probe, ScrapeConfig
+* Replica
+* Shard
+* Storage
```

Comment:
@ArthurSens mentioned that we may be able to leverage [CEL](https://kubernetes.io/docs/reference/using-api/cel/) to prevent users setting unsupported fields in daemonset mode. But we also agreed that it's not required for the MVP. As a first approach, the operator can fail the reconciliation in such cases though.

Reply:
It would be nice to mention this in an `Alternatives` section, where we acknowledge different approaches but explain why we decided to move forward with the proposed approach.

Reply:
This doesn't seem like an alternative, but an enhancement. AFAIU alternatives mean what we considered, but decided not do do. Since we may do it after the MVP, should we place it in `Follow-ups` instead?

Reply:
@simonpasquier @kakkoyun @ArthurSens Could you have a look at the current PrometheusAgent CRD to see if there're any other fields not suitable for DaemonSet, besides the ones I listed here?

Reply:
Looks good to me.